src/VisualStudio/Core/Def/Implementation/ProjectSystem/CPS/IWorkspaceProjectContext.cs

```diff
@@ -36,6 +39,10 @@ internal interface IWorkspaceProjectContext : IDisposable
         void RemoveSourceFile(string filePath);
         void AddAdditionalFile(string filePath, bool isInCurrentContext = true);
         void RemoveAdditionalFile(string filePath);
+        void AddSourceTextContainer(SourceTextContainer textContainer, string fullPath, IEnumerable<string> folderNames = null, SourceCodeKind sourceCodeKind = SourceCodeKind.Regular);
+        void RemoveSourceTextContainer(SourceTextContainer textContainer);
+        Task AddNonSourceFileAsync(string filePath, bool isInCurrentContext = true, IEnumerable<string> folderNames = null, CancellationToken cancellationToken = default);
+        Task RemoveNonSourceFileAsync(string fullPath, CancellationToken cancellationToken);
```

Comment:
Looks these these are the two new APIs yes? @davkean

Some questions:

Do these need to be async? I don't remember that as something we discussed
Do we still need `AddSourceTextContainer`/`RemoveSourceTextContainer`?

Reply:
@jasonmalinowski asked me to make it Async. he said, he is planing to change all existing ones to Async. SouceTextContainers are removed.

Reply:
OK, just checking. As long as @davkean and @jasonmalinowski are in agreement, then this API seems like what I expect.

Reply:
Please make them async all at once, not piecemeal - we call these without async context.

Reply:
let me know what you want on the async side.